overlord/devicestate/devicemgr.go

```diff
@@ -797,15 +797,52 @@ func systemFromSeed(label string, current *seededSystem) (*System, error) {
 		Brand:   brand,
 		Actions: defaultSystemActions,
 	}
-	if current != nil && isCurrentSystem(current, &system) {
-		system.Current = true
-		system.Actions = currentSystemActions
-	}
 	return &system, nil
 }
 
 var ErrNoSystems = errors.New("no systems seeds")
 
+func currentSystemForMode(st *state.State, mode string) (*seededSystem, error) {
+	switch mode {
+	case "run":
+		return currentSeedSystem(st)
+	case "install":
+		// there is no current system for install mode
+		return nil, nil
+	case "recover":
+		// recover mode uses modeenv for reference
+		return seededSystemFromModeenv()
+	}
+	return nil, fmt.Errorf("internal error: cannot identify current system for unsupported mode %q", mode)
+}
+
+func seededSystemFromModeenv() (*seededSystem, error) {
+	modeEnv, err := maybeReadModeenv()
+	if err != nil {
+		return nil, err
+	}
+	if modeEnv == nil {
+		return nil, fmt.Errorf("internal error: modeenv does not exist")
+	}
+	if modeEnv.RecoverySystem == "" {
+		return nil, fmt.Errorf("internal error: recovery system is unset")
+	}
+
+	system, err := systemFromSeed(modeEnv.RecoverySystem, nil)
+	if err != nil {
+		return nil, err
+	}
+	seededSys := &seededSystem{
+		System:    modeEnv.RecoverySystem,
+		Model:     system.Model.Model(),
+		BrandID:   system.Model.BrandID(),
+		Revision:  system.Model.Revision(),
+		Timestamp: system.Model.Timestamp(),
+		// SeedTime is intentionally left unset
+	}
+	return seededSys, nil
+}
+
 func currentSeedSystem(st *state.State) (*seededSystem, error) {
```

Comment:
as we discussed I think we should split these helpers out,  I would create a systems.go file with everything from the struct definitions for SystemAction and System down to this helper.

Reply:
I'll push the cleanup in a follow up